src/isl/isl_type_reference.rs

```diff
@@ -5,18 +5,39 @@ use crate::result::{
     invalid_schema_error, invalid_schema_error_raw, unresolvable_schema_error, IonSchemaResult,
 };
 use crate::system::{PendingTypes, TypeId, TypeStore};
-use crate::types::TypeDefinitionImpl;
+use crate::types::{CoreISLTypeDefinition, TypeDefinitionImpl};
 use ion_rs::value::owned::OwnedElement;
+use ion_rs::value::reader::{element_reader, ElementReader};
 use ion_rs::value::{Element, Struct, SymbolToken};
 use ion_rs::IonType;
+use std::collections::HashMap;
+
+/// Represents all [Core ISL types]
+/// This also includes: lob, number, text, any
+/// [Core ISL types]: https://amzn.github.io/ion-schema/docs/spec.html#core-types
+#[derive(Debug, Clone, PartialEq)]
+pub enum CoreIslType {
+    IonType(IonType),
+    Other(IslTypeImpl),
+}
```

Comment:
question—Is `IonType` for types like `$string` and `$blob` and `Other` is for `string`, `blob`, `any`, `$any`, etc?

Reply:
As per this implementation:
```
IonType - string, blob, clob, int, decimal, float (no '$' prefixed ion types)
Other - any, lob, number, text (no '$' prefixed ion types)
```

Reply:
Renamed all builtin types:
   - CoreIslTypeDeifnition --> BuitlInTypeDefinition
   - Atomic(IonType, bool) --> all ion types and all '$' prefixed ion types where bool represents whether they allow null or not
   **note: This can be changed to just be `Atomic(IonType)`. Once `nullable` is implemented '$' prefixed ion types can be added to `Derived` with `type:: { name: '$string', type: nullable:: string }`**
   - Derived(TypeDefinitionImpl) --> all derived ISL types like any, number, text, lob, $any, $number, $text, $lob

Reply:
I think this approach is fine, except for one thing (which is non-blocking for this PR).

> Once nullable is implemented '$' prefixed ion types can be added to Derived with `type:: { name: '$string', type: nullable:: string }`

Actually, `nullable::string` is not equivalent to `$string`—`nullable::string` would allow `null.null` and `null.string`, whereas `$string` would allow only `null.string`. Also, [`nullable` is a bit messy](https://github.com/amzn/ion-schema/issues/38), and I don't think it's wise to use it as part of defining some of the types that are part of the Ion Schema language.

On further thought, you also cannot define `$string` as `{ one_of: [ string, { valid_values: [null.string] }] }` because of the implicit `type: any`.

That leaves us with 3 options.
1.  `$string` is the Atomic type, and then we derive `string` as `{ type: $string, not: { valid_values: [null.string] } }`
2. Both `$string` and `string` are Atomic types
3. We create a new Atomic type `any_null` (or something like that), so that we can get around the implicit `type: any` problem I mentioned above.

But again, this is something that does not need to be decided right now.

Reply:
I think for now we have option#2 implemented, which seems fine. Only if we have a strong preference for `Atomic` types to be atomic meaning only have `string` and not `$string`, we can go for option#1 once we have `valid_values`.